### PR TITLE
 Fix: Inclusion of DateTime in a Date range and vice versa

### LINKED ISSUE #36175: DateTime, Date comparison doesn't work as expected
### Steps to reproduce

```ruby
may1 = Date.today.beginning_of_month # => Wed, 01 May 2019
may31 = Date.today.end_of_month # => Fri, 31 May 2019

(may1..may31).include? Date.today # => True
(may1..may31).include? DateTime.now # => False
```

### Expected behavior
`(may1..may31).include? DateTime.now` should return `True` for DateTime's in the same month

### Actual behavior
It returns `False`

### System configuration
**Rails version**: Rails 5.2.3

**Ruby version**: ruby 2.6.0p0 (2018-12-25 revision 66547) [x86_64-linux]

### PR DESCRIPTION
**Pull request by vishaltelangre**
Fixes https://github.com/rails/rails/issues/36175 which exists for Rails 5.2.3 and Rails 6, too.

With this fix, the range inclusion predicate method works as intended.

```ruby
(Date.today.beginning_of_month..Date.today.end_of_month).include?(Date.today) # => true
(Date.today.beginning_of_month..Date.today.end_of_month).include?(DateTime.now) # => true
(DateTime.now.beginning_of_month..DateTime.now.end_of_month).include?(Date.today) # => true
(DateTime.now.beginning_of_month..DateTime.now.end_of_month).include?(DateTime.now) # => true
```

### CODE REVIEW

File: activesupport/test/core_ext/range_ext_test.rb
Comment (st0012) [inline activesupport/test/core_ext/range_ext_test.rb]: Should also add test cases for `Time` inclusion. Like:

```ruby
(Date.yesterday..Date.tomorrow).include?(Time.now)
(Date.yesterday..Date.tomorrow).include?(Time.zone.now)
```

**Review by st0012 — CHANGES_REQUESTED** [review]

File: activesupport/test/core_ext/range_ext_test.rb
Comment (vishal-vineti) [reply activesupport/test/core_ext/range_ext_test.rb]: Added tests for this.

Comment (vishaltelangre) [reply activesupport/test/core_ext/range_ext_test.rb]: Added tests for these scenarios, too.

**Comment by st0012** [comment]
@vishaltelangre I've opened an issue report on Ruby to check if the change should be done at Ruby level instead: https://bugs.ruby-lang.org/issues/16440

In the meantime, I think the PR looks good. Can you squash your commits so once they make the call it can easily be merged?

**Comment by zverok** [comment]
[Range#include?](https://ruby-doc.org/core-2.6.5/Range.html#method-i-include-3F) in Ruby checks if the argument is one of the _members_ of the range-as-a-sequence, e.g. if `Range#to_a` will include it.
To check whether the argument is in the space between `begin` and `end`, [Range#cover?](https://ruby-doc.org/core-2.6.5/Range.html#method-i-cover-3F) should be used.

I believe this kind of "edge case hacks", even if they look good on one particular example, are in fact harmful (for AS users) because of muddying the underlying Ruby semantics.

**Comment by vishaltelangre** [comment]
@st0012 Is this PR still a candidate to be merged?

**Comment by st0012** [comment]
@vishaltelangre unfortunately, it's not my call 😬. 
Based on what I see now, it probably won't be merged, unfortunately. If they decide the behavior is valid, they will implement it at the Ruby level instead of Rails. And if they don't think this is expected behavior, Rails maintainers probably won't want to support it in Rails either, unless there's a strong argument for having this feature at the Rails level.

**Comment by vishaltelangre** [comment]
@st0012 Sounds good to me. I will leave this PR as is. It will be automatically closed by @rails-bot once it becomes stale.

**Comment by st0012** [comment]
@vishaltelangre my bug report in Ruby has been rejected

> include? should only be used if you want to check the argument is one of the members of the range (i.e. included in the array returned by to_a)

So I guess we should follow the same standard in Rails.

**Comment by vishaltelangre** [comment]
@st0012 Thanks for the update. In that case, should we close this fix as well?

**Comment by rafaelfranca** [comment]
I agree with @zverok here. Does `cover?` work for the examples you gave in the description?

File: activesupport/lib/active_support/core_ext/range/include_date_time.rb
Comment (vishaltelangre) [inline activesupport/lib/active_support/core_ext/range/include_date_time.rb]: @rafaelfranca That's what this PR does. We rely on `#cover?` here. This fix is inspired by this extension https://github.com/rails/rails/blob/870f087825/activesupport/lib/active_support/core_ext/range/include_time_with_zone.rb which is already present in the upstream.

Comment (rafaelfranca) [reply activesupport/lib/active_support/core_ext/range/include_date_time.rb]: I'm more inclined to remove that extension. `include?` should not work for values that `Range.to_a` would not return.

Can we deprecate that extension and tell people to use `cover?` instead?

Comment (vishaltelangre) [reply activesupport/lib/active_support/core_ext/range/include_date_time.rb]: @rafaelfranca Sure. I will make another PR to deprecate the `activesupport/lib/active_support/core_ext/range/include_time_with_zone.rb` extension.

**Comment by vishaltelangre** [comment]
Closing this PR since it is encouraged to use the `#cover?` method to check the inclusion of DateTime in a Date range and vice versa instead of using the `#include?` method to do the same.

**Pull request closed without merging**
